src/client/src/Client.c

```diff
@@ -20,6 +20,84 @@ extern UINT32 CLIENT_STATE_MACHINE_STATE_COUNT;
  */
 PKinesisVideoClient gKinesisVideoClient = NULL;
 
+
+/**
+ *
+ * @param timerId - timerId for timer
+ * @param currentTime - the current time when the call back was fired
+ * @param customData - pKinesisVideoClient, contains the streams and interval info
+ * @return
+ */
+STATUS checkIntermittentProducerCallback(UINT32 timerId, UINT64 currentTime, UINT64 customData)
+{
+    ENTERS();
+    UNUSED_PARAM(timerId);
+    STATUS retStatus = STATUS_SUCCESS;
+    PKinesisVideoClient pKinesisVideoClient = (PKinesisVideoClient) customData;
+    UINT32 i;
+    BOOL clientStreamsListLocked = FALSE, streamLocked = FALSE;
+    PKinesisVideoStream pCurrStream = NULL;
+    Frame eofr = EOFR_FRAME_INITIALIZER;
+
+    CHK(pKinesisVideoClient, STATUS_NULL_ARG);
+
+    pKinesisVideoClient->timerCallbackInvocationCount++;
+
+    // Lock the client streams list lock
+    pKinesisVideoClient->clientCallbacks.lockMutexFn(pKinesisVideoClient->clientCallbacks.customData,
+                                                     pKinesisVideoClient->base.streamListLock);
+    clientStreamsListLocked = TRUE;
+    for (i = 0; i < pKinesisVideoClient->deviceInfo.streamCount; i++) {
+        if (NULL != pKinesisVideoClient->streams[i]) {
+            pCurrStream = pKinesisVideoClient->streams[i];
+            // Validate stream isn't being shut down or is closed
+            if (!(pCurrStream->streamStopped || pCurrStream->streamClosed)) {
+                // Lock the Stream
+                pKinesisVideoClient->clientCallbacks.lockMutexFn(pKinesisVideoClient->clientCallbacks.customData,
+                                                                 pCurrStream->base.lock);
+                streamLocked = TRUE;
+                // Check if last PutFrame is older than max timeout, if so, send EoFR, if not, do nothing
+                // Ignoring currentTime it COULD be smaller than pCurrStream->lastPutFrametimestamp
+                // Due to this method entering but waiting on stream lock from putFrame call
+                if (IS_VALID_TIMESTAMP(pCurrStream->lastPutFrameTimestamp) &&
+                    (GETTIME() - pCurrStream->lastPutFrameTimestamp) > INTERMITTENT_PRODUCER_MAX_TIMEOUT) {

```

Comment:
GETTIME can be expensive. Store the value of GETTIME before the loop and use in the iteration

Reply:
It will deadlock that way, it happened during one of the existing tests.  We entered call back but another thread (in putFrame) had the stream lock and was updating `lastPutFrameTimestamp`, by the time it released the lock the saved GETTIME() value is smaller than `pCurrStream->lastPutFrameTimestamp` resulting in `(GETTIME() - pCurrStream->lastPutFrameTimestamp)` those are UINT64 but we're trying to subtract a larger value from a smaller one -- same reason why as stated in the commend I couldn't just use `currentTime` which is passed into the function.

There are a few alternatives here I can just explicitly check if the saved currTime value (or the result of GETTIME() before the loop/lock acquire) is < pCurrStream->lastPutFrameTimestamp and if it is we know there's nothing to do so we can bail. I'll make that change but just wanted to bring up the thinking here.

Reply:
Which lock is being contended? If you refer to pClient->mTestClientMutex then this is a fast/prompt operation. Why does this deadlock?